flashlight/fl/nn/modules/Module.h

```diff
@@ -122,6 +141,13 @@ class Module {
    */
   std::vector<Variable> operator()(const std::vector<Variable>& inputs);
 
+  /**
+   * Clone the module via deep copy of its parameters and members.
+   *
+   * @return A unique pointer of the cloned module.
+   */
+  virtual std::unique_ptr<Module> clone() const = 0;
```

Comment:
This would only be required if the module is a base pointer, otherwise value copy semantics can be used. Where a `cloneModule()` function would be useful is when cloning via calling the base `Module` virtual clone method from a smart pointer. So I could add something like this:

```cpp
template <typename T, typename D, template <typename, typename> class SmartPtr>
SmartPtr<T, D> cloneModule(const SmartPtr<Module, D>& module) {
  auto copy = module->clone();
  return SmartPtr<T, D>(dynamic_cast<T*>(copy.release()), std::move(copy.get_deleter()));
}
```
Which would be used like:
```cpp
std::unique_ptr<Module> modulePtr = std::make_unique<Linear>(10, 20);
std::unique_ptr<Linear> clonedPtr = cloneModule<Linear>(modulePtr);
// or
std::shared_ptr<Module> modulePtr = std::make_shared<Linear>(10, 20);
std::shared_ptr<Linear> clonedPtr = cloneModule<Linear>(modulePtr);
```

Reply:
This implementation with smart pointers makes sense.

As a corollary to this, though, (and playing devil's advocate for a moment) what if we considered using a `cloneModule` function that had overloads for most container types (in addition to derived types of `Module`), and removed the `clone()` function on the `Module` interface entirely, using that instead?

Reply:
Removing `clone()` from `Module` and replacing with a free standing `cloneModule` function with overrides for all module implementations will not solve the type erasure issue when a pointer to the base `Module` is used (for example in `Container` or in `WeightNorm`). The derived type is unknown at compile time, hence the correct copy constructor can't be called:

```cpp
std::unique_ptr<Module> linearPtr = std::make_unique<Linear>(10, 20);
Linear linearCopy = *linearPtr; // **compile time error** unless the module is cast to its derived type, but we may not know the type at compile time
```
If the type is known at compile time:
```cpp
std::unique_ptr<Linear> linearPtr = std::make_unique<Linear>(10, 20);
Linear linearCopy = *linearPtr; // Copy via copy ctor as the derived type is known at compile time
std::unique_ptr<Linear> clonedPtr = cloneModule<Linear>(linearPtr); // or via the proposed cloneModule function
```

Reply:
I think there may be some confusion here (or perhaps I'm misunderstanding?). The `cloneModule<T>()` function does not provide a compile time-guaranteed-safe method of cloning, its purely a convenience function. The type is erased at compile time and run time, hence it is possible for the user to specify the wrong type when casting and because the actual type is only known at run time, at the very least a dynamic cast must be used. Otherwise using the wrong type leads to UB.

> An alternative idea might be to make Module::clone protected and cloneModule a friend, though the templates are tricky.

I dont believe this is possible as the template parameter for `cloneModule<T>` would be unknown to `Module`.

> Orthogonally -- could we not have a cloneModule implementation that operates on stack allocated modules as well or takes a const T& where we static assert std::is_base_of<Module, T>::value?

If a module is stack allocated its type must be known, hence the copy constructor can be used instead of using `Module:clone()`. `Module:clone()` is only necessary when the type has been erased or if it is desired to obtain a type erased clone.

I suspect the real problem here is that there is no reflection in C++ (both compile and run time). So common functionality used in python (e.g. metaclasses) is not possible at this stage. Compile time reflection may be added in C++26, but at the rate its going its unlikely to land.

Reply:
Good points here, and of course safety isn't guaranteed given that there's a cast, but that's a constraint of the return type of `clone()` being `Module`.

I see there still being merits for `cloneModule` -- low hanging fruit it saves the user a cast from `Module` (returned by `clone`) back to the base type. It's the difference between:
```cpp
std::unique_ptr<Linear> linear = ...;
std::unique_ptr<Linear> linear2 = cloneModule(linear);
// versus
std::unique_ptr<Linear> linear2 = std::dynamic_pointer_cast<Linear>(linear->clone());
```
with the benefit that the correctness of the cast is guaranteed (though the cast is still required, of course), meaning `static_pointer_cast` can be used, with assurances. This saves a bit of overhead/removes some ugliness.

> dont believe this is possible as the template parameter for cloneModule<T> would be unknown to Module

Template friends mean that any that any T for ` cloneModule<T>` that's a friend of `Module` would be accepted, so this approach is viable ([if my understanding is correct](https://en.cppreference.com/w/cpp/language/friend#Template-friends)). If we made `clone` protected, in this case, there would be no dynamic casts present in end-to-end usage of the API, which in my view would be much less error prone.

Reply:
> I see there still being merits for cloneModule -- low hanging fruit it saves the user a cast from Module

Yep, completely agree with this. My perspective is that its purely a convenience function.

> Template friends mean that any that any T for cloneModule<T> that's a friend of Module would be accepted, so this approach is viable

Ah I misunderstood what you were trying to achieve regarding templated friend functions, you are indeed correct that its viable! Though, perhaps I'm still misunderstanding what you are trying to achieve, as I'm not sure what making `cloneModule` a friend would do that using the public interface of `Module` wouldn't? Unless you are thinking of setting `train_` for the cloned/copied module?

> If we made clone protected, in this case, there would be no dynamic casts present in end-to-end usage of the API, which in my view would be much less error prone.

I'm not following this statement as based on my understanding if the derived type is known at compile time, then the copy constructor can be called rather than calling `cloneModule()` or `Module::clone()`. If the derived type is unknown at compile time, then we must necessarily use dynamic_cast to ensure the module is being cast to the correct type.

Thinking about this more and in regards to an earlier comment about using `cloneModule` as a single source for cloning, I don't believe it's possible as the copy constructor for derived modules can also be called. I'm not sure if there is a reasonable approach to enforce making the copy constructor of derived modules protected without completely changing the design of this PR and a lot of Flashlight itself (i.e. using the macro approach that PyTorch C++ API implements).

Reply:
After playing with this myself finally, I agree with your perspective. To summarize:
1. We need `clone()` no matter what because we can't have virtual ctors (`clone` is literally the idiom, of course)
2. It's necessary that we express that an abstract `Module` instance _is_ copyable if the derived type isn't known and provide a path to doing so
3. `cloneModule` is only useful insofar as it works on abstract `Module`s. We don't know the derived type, so we have a dynamic cast no matter what.

Overall, since `cloneModule` really just hides `dynamic_cast`, I'd actually rather we remove it (I was wrong!) and have the user do dicey things with casting instead of condoning it in an API. I'm sorry to make you spin your wheels on this; thanks for sticking with me on it :)

Reply:
Under the right circumstance `cloneModule` could be valuable to have, but I do agree that it hides dynamic_cast and does give users more tools to shoot themselves in the foot. If a user does find it useful to have, its such a small function, they can easily implement it themselves.